zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/launcher/SparkInterpreterLauncher.java

```diff
@@ -0,0 +1,186 @@
+      env.put("HADOOP_CONF_DIR", System.getenv("HADOOP_CONF_DIR"));
+    }
+    LOGGER.debug("buildEnvFromProperties: " + env);
+    return env;
+
+  }
+
+  private boolean isSparkConf(String key, String value) {
+    return !StringUtils.isEmpty(key) && key.startsWith("spark.") && !StringUtils.isEmpty(value);
+  }
+
+  private void setupPropertiesForPySpark(Properties sparkProperties) {
+    if (isYarnMode()) {
+      sparkProperties.setProperty("spark.yarn.isPython", "true");
+    }
+  }
+
+  private void mergeSparkProperty(Properties sparkProperties, String propertyName,
+                                  String propertyValue) {
+    if (sparkProperties.containsKey(propertyName)) {
+      String oldPropertyValue = sparkProperties.getProperty(propertyName);
+      sparkProperties.setProperty(propertyName, oldPropertyValue + "," + propertyValue);
+    } else {
+      sparkProperties.setProperty(propertyName, propertyValue);
+    }
+  }
+
+  private void setupPropertiesForSparkR(Properties sparkProperties,
+                                        String sparkHome) {
+    File sparkRBasePath = null;
+    if (sparkHome == null) {
+      if (!getSparkMaster(properties).startsWith("local")) {
+        throw new RuntimeException("SPARK_HOME is not specified in interpreter-setting" +
+            " for non-local mode, if you specify it in zeppelin-env.sh, please move that into " +
+            " interpreter setting");
+      }
+      String zeppelinHome = zConf.getString(ZeppelinConfiguration.ConfVars.ZEPPELIN_HOME);
+      sparkRBasePath = new File(zeppelinHome,
+          "interpreter" + File.separator + "spark" + File.separator + "R");
+    } else {
+      sparkRBasePath = new File(sparkHome, "R" + File.separator + "lib");
+    }
+
+    File sparkRPath = new File(sparkRBasePath, "sparkr.zip");
+    if (sparkRPath.exists() && sparkRPath.isFile()) {
+      mergeSparkProperty(sparkProperties, "spark.yarn.dist.archives", sparkRPath.getAbsolutePath());
+    } else {
+      LOGGER.warn("sparkr.zip is not found, SparkR may not work.");
+    }
+  }
+
+  /**
+   * Order to look for spark master
+   * 1. master in interpreter setting
+   * 2. spark.master interpreter setting
+   * 3. SPARK_HOME in zeppelin-env.sh
+   * 4. use local[*]
+   * @param properties
+   * @return
+   */
+  private String getSparkMaster(Properties properties) {
+    String master = properties.getProperty("master");
+    if (master == null) {
+      master = properties.getProperty("spark.master");
+      if (master == null) {
+        master = System.getenv("SPARK_HOME");
+      }
+      if (master == null) {
+        master = "local[*]";
+      }
+    }
+    return master;
+  }
+
+  private String getDeployMode() {
+    String master = getSparkMaster(properties);
+    if (master.equals("yarn-client")) {
+      return "client";
+    } else if (master.equals("yarn-cluster")) {
+      return "cluster";
+    } else if (master.startsWith("local")) {
+      return "client";
```

Comment:
why is "local" == "client"?

Reply:
isn't this never going to hit? isYarnMode is true before calling this method

Reply:
there's only 2 valid value for `spark.submit.deployMode` (client or cluster). It is `client` for local mode IIUC.